BUG: Fix closed surface disappearing when segment is removed

Sunderlandkyl opened the pull request:
When a segment that is in the same layer as other segments is removed from a vtkSegmentation, it causes the shared labelmap to be modified. This triggered a SourceRepresentationModified on the remaining segments, which invalidated non-source representations. Fixed by calling SetSourceRepresentationModifiedEnabled(false) before editing the labelmap, and restoring the original value after.

Fixes #7220

lassoan reviewed (commented):
Thank you for the quick fix. Are there any usages of `SeparateSegmentLabelmap` that would need to be similarly fixed up? Even better, could we call `SetSourceRepresentationModifiedEnabled` inside `SeparateSegmentLabelmap` (or that could suppress update events and cause issues)?

Sunderlandkyl commented:
I just tested it, and there is another use that causes the same issue. Trying to paint over another segment with "Allow overlap" will cause the 3d surface to disappear. It makes sense to put SetSourceRepresentationModifiedEnabled inside SeparateSegmentLabelmap.

jcfr commented:
I did some testing to identify what introduced the regression:

| revision | hash | status |
|--|--|--|
| 31503 | 7f5f83e | :white_check_mark: |
| 31739 | 9ae54bd | :white_check_mark: |
| 31800 | 7195f78 | :white_check_mark: |
| 31807 | d7c9a0d | :white_check_mark: |
| 31810 | 1bd1dcf | :white_check_mark: |
| 31813 | d12bcab | :x: |
| 31843 | 5a5de07 | :x: |

The regression was likely introduced in d12bcabc2b777512ae310c2117fd2195ab0cccfb (`ENH: Update VTK from 9.1.20220125 to 9.2.20230607`) integrated through:
* https://github.com/Slicer/Slicer/pull/7010

![image](https://github.com/Slicer/Slicer/assets/219043/187b7576-b348-4cc2-92c1-64f74683da40)

Sunderlandkyl commented:
Thanks @jcfr. It looks like ShallowCopy didn't trigger a ModifiedEvent with the previous version of VTK.

@lassoan I've moved SetSourceRepresentationModifiedEnabled inside SeparateSegmentLabelmap.
I've also removed InvokeEvent "SourceRepresentationModified" and "ContainedRepresentationNamesModified" as I don't think they should be triggered manually by the function.

lassoan commented on Libs/vtkSegmentationCore/vtkSegmentation.cxx:
```diff
@@ -1573,13 +1573,24 @@ void vtkSegmentation::SeparateSegmentLabelmap(std::string segmentId)
+    // We were blocking onSourceRepresentationModified, however the source labelmap was modified so we need to manually invoke the event.
+    this->InvokeEvent(vtkSegmentation::SourceRepresentationModified, labelmap);
+    }
+  this->InvokeEvent(vtkSegmentation::RepresentationModified, labelmap);
```
Could you please check if these 3 events (vtkSegmentation::SourceRepresentationModified, vtkSegmentation::RepresentationModified, this->Modified()) do not result in performing any expensive operation (e.g., regeneration of closed surface representation) multiple times? (most probably not, just to make sure)

Sunderlandkyl replied:
Summary of event observations:
- SourceRepresentationModified:
  - updateWidgetFromMRML in qSlicerSegmentationsModuleWidget
  - updateSliceRotateWarningButtonVisibility in qMRMLSegmentEditorWidget.
  - populateRepresentationsList in qMRMLSegmentationRepresentationsListView.
  - Segmentation history
  - Update storage node supported write file types.
  - Propagate event through segmentation node.

- RepresentationModified:
  - Displayable managers.
  - Update StorableModifiedTime.
  - Propagate event through segmentation node..

- ModifiedEvent:
  - Not observed anywhere

lassoan commented on Libs/vtkSegmentationCore/vtkSegmentation.cxx:
```diff
@@ -1573,13 +1573,24 @@ void vtkSegmentation::SeparateSegmentLabelmap(std::string segmentId)
     thresholdErase->ReplaceOutOff();
     thresholdErase->Update();
+
+    // Although the source labelmap will be modified, onSourceRepresentationModified would invalidate the non-source representations.
```
vtkMRMLSegmentationNode::OnSourceRepresentationModified() or vtkSegmentation::OnSourceRepresentationModified()?

Sunderlandkyl replied:
vtkSegmentation::OnSourceRepresentationModified()

lassoan commented on Libs/vtkSegmentationCore/vtkSegmentation.cxx:
```diff
@@ -1573,13 +1573,24 @@ void vtkSegmentation::SeparateSegmentLabelmap(std::string segmentId)
     thresholdErase->ReplaceOutOff();
     thresholdErase->Update();
+
+    // Although the source labelmap will be modified, onSourceRepresentationModified would invalidate the non-source representations.
```
```suggestion
    // Although the source labelmap will be modified, onSourceRepresentationModified would unnecessarily invalidate the non-source representations
    // (because separating a labelmap into a new layer does not change the segment's shape).
```

lassoan commented on Libs/vtkSegmentationCore/vtkSegmentation.cxx:
```diff
@@ -1573,13 +1573,24 @@ void vtkSegmentation::SeparateSegmentLabelmap(std::string segmentId)
     thresholdErase->Update();
+
+    // Although the source labelmap will be modified, onSourceRepresentationModified would invalidate the non-source representations.
+    // This prevents onSourceRepresentationModified from being called.
```
```suggestion
    // This prevents onSourceRepresentationModified from being called. We will later invoke events to let observers know about the change. 
```

lassoan commented on Libs/vtkSegmentationCore/vtkSegmentation.cxx:
```diff
@@ -1573,13 +1573,24 @@ void vtkSegmentation::SeparateSegmentLabelmap(std::string segmentId)
     thresholdErase->SetInValue(0);
     thresholdErase->ReplaceOutOff();
     thresholdErase->Update();
+
```
I've just noticed above in line 555: `if (labelmap)...` - why don't we return immediately if `labelmap == nullptr` ?

lassoan reviewed (commented):
Thank you, this looks good! I've just added a few (mostly trivial) comments and suggestions.

Sunderlandkyl commented:
Thanks @lassoan, changes made!

lassoan reviewed (approved):
Thank you, it all looks good.

jcfr commented:
> ContainedRepresentationNamesModified is still removed

For reference, this was first introduced in fcb40b0738c65ab418731a1a34292f3c85b2133a (`ENH: Add support for shared binary labelmap segmentations`)
